client-local/src/main/java/com/cloudera/livy/client/local/LocalClient.java

```diff
@@ -249,75 +252,12 @@ public void run() {
       // SparkSubmit will take care of that for us.
       String master = conf.get("spark.master");
       Preconditions.checkArgument(master != null, "spark.master is not defined.");
-
-      List<String> argv = Lists.newArrayList();
-
-      if (sparkHome != null) {
-        argv.add(new File(sparkHome, "bin/spark-submit").getAbsolutePath());
-      } else {
-        LOG.info("No spark.home provided, calling SparkSubmit directly.");
-        argv.add(new File(System.getProperty("java.home"), "bin/java").getAbsolutePath());
-
-        if (master.startsWith("local") ||
-              master.startsWith("mesos") ||
-              master.endsWith("-client") ||
-              master.startsWith("spark")) {
-          String mem = conf.get("spark.driver.memory");
-          if (mem != null) {
-            argv.add("-Xms" + mem);
-            argv.add("-Xmx" + mem);
-          }
-
-          String cp = conf.get("spark.driver.extraClassPath");
-          if (cp != null) {
-            argv.add("-classpath");
-            argv.add(cp);
-          }
-
-          String libPath = conf.get("spark.driver.extraLibPath");
-          if (libPath != null) {
-            argv.add("-Djava.library.path=" + libPath);
-          }
-
-          String extra = conf.get(DRIVER_OPTS_KEY);
-          if (extra != null) {
-            for (String opt : extra.split("[ ]")) {
-              if (!opt.trim().isEmpty()) {
-                argv.add(opt.trim());
-              }
-            }
-          }
-        }
-
-        argv.add("org.apache.spark.deploy.SparkSubmit");
-      }
-
-      if (master.equals("yarn-cluster")) {
-        String executorCores = conf.get("spark.executor.cores");
-        if (executorCores != null) {
-          argv.add("--executor-cores");
-          argv.add(executorCores);
-        }
-
-        String executorMemory = conf.get("spark.executor.memory");
-        if (executorMemory != null) {
-          argv.add("--executor-memory");
-          argv.add(executorMemory);
-        }
-
-        String numOfExecutors = conf.get("spark.executor.instances");
-        if (numOfExecutors != null) {
-          argv.add("--num-executors");
-          argv.add(numOfExecutors);
-        }
-      }
-
-      argv.add("--properties-file");
-      argv.add(sparkConf.getAbsolutePath());
-      argv.add("--class");
-      argv.add(RemoteDriver.class.getName());
+      launcher.setMaster(master);
+      launcher.setPropertiesFile(sparkConf.getAbsolutePath());
+      launcher.setMainClass(RemoteDriver.class.getName());
 
       String jar = "spark-internal";
+      launcher.setAppResource(jar);
       String livyJars = conf.get(LIVY_JARS);
```

Comment:
Not your fault, but this whole block (down to L282) should be moved before L247, where the configuration is written out. Otherwise it won't take effect in certain cases...